### DISCUSSION TITLE
当右上角X关闭窗口时报错(An error occurs when the upper right corner x closes the window)

### DISCUSSION THREAD

**Discussion by RONG-GUI** [PyQt5/PyQt#151] [Q&A] [answered]
当右上角X关闭窗口时报错(An error occurs when the upper right corner x closes the window)：
"Fatal Python error: _enter_buffered_busy: could not acquire lock for <_io.BufferedReader name='<stdin>'> at interpreter shutdown, possibly due to daemon threads
Python runtime state: finalizing (tstate=000001CA6F3710E0)

Current thread 0x00001c40 (most recent call first):
no Python frame"

尝试重写 QtWidgets.QMainWindow的closeEvent也是一样。
Try to rewrite closeEvent, but the result is the same.

----code------
`import sys
from PyQt6 import QtWidgets
from PyQt6.QtCore import QThread
from PyQt6.QtCore import pyqtSignal
import PD  # PyUic生成的UI文件
 
 
class Windows(QtWidgets.QMainWindow, PD.Ui_MainWindow):
 
    def __init__(self):
        super(Windows, self).__init__()
        self.setupUi(self)
        self.add_text = AddText()
        # self.add_text.setDaemon(True)  # QThread没有setDaemon
        self.add_text.start()
        self.add_text.signal.connect(self.add_box)
 
    def add_box(self, text):
        pass
 
 
class AddText(QThread):
 
    signal = pyqtSignal(list)
 
    def __init__(self):
        """子线程，用于调试文本"""
        super(AddText, self).__init__()
        self.txt = list()

     def run(self):
        self.input_txt()

    def input_txt(self):
        try:
            while True:
                val = input("请输入：")
                self.txt.append(val)
                self.signal.emit(self.txt)
        except Exception as e:
            pass
 
 
if __name__ == '__main__':
    app = QtWidgets.QApplication(sys.argv)
    MainWindow = QtWidgets.QMainWindow()
    ui = Windows()
    ui.setupUi(MainWindow)
    MainWindow.show()
    sys.exit(app.exec())`

**Comment by RONG-GUI** [comment, marked as answer]
已解决

**Reply by feelinglessly** [reply]
怎么解决的啊

**Comment by 892768447** [comment]
> 怎么解决的啊

1. 线程里不能直接操作UI。
2. 主线程关闭时等待线程退出（具体可以参考下这个例子 https://github.com/PyQt5/PyQt/blob/master/QThread/InheritQThread.py#L27）
